sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnSignature.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import static org.apache.beam.sdk.transforms.DoFn.FinishBundle;
+import static org.apache.beam.sdk.transforms.DoFn.ProcessElement;
+import static org.apache.beam.sdk.transforms.DoFn.StartBundle;
+import static org.apache.beam.sdk.transforms.DoFnReflector.format;
+import static org.apache.beam.sdk.transforms.DoFnReflector.formatType;
+
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+
+import com.google.common.reflect.TypeParameter;
+import com.google.common.reflect.TypeToken;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Describes the signature of a {@link DoFn}, in particular, which features it uses, which extra
+ * context it requires, types of the input and output elements, etc.
+ *
+ * <p>See <a href="https://s.apache.org/a-new-dofn">A new DoFn</a>.
+ */
+class DoFnSignature {
+  public final TypeToken<?> inputT;
+  public final TypeToken<?> outputT;
+  public final ProcessElementMethod processElement;
+  public final BundleMethod startBundle;
+  public final BundleMethod finishBundle;
+
+  private DoFnSignature(
+      TypeToken<?> inputT,
+      TypeToken<?> outputT,
+      ProcessElementMethod processElement,
+      BundleMethod startBundle,
+      BundleMethod finishBundle) {
+    this.inputT = inputT;
+    this.outputT = outputT;
+    this.processElement = processElement;
+    this.startBundle = startBundle;
+    this.finishBundle = finishBundle;
+  }
+
+  /** Describes a {@link DoFn.ProcessElement} method. */
+  public static class ProcessElementMethod {
+    /** The relevant method in the user's class. */
+    public final Method targetMethod;
+    /** 0-based index of the {@link BoundedWindow} parameter, or -1 if it's not used. */
+    public final int boundedWindowParamIndex;
+    /** 0-based index of the {@link DoFn.InputProvider} parameter, or -1 if it's not used. */
+    public final int inputProviderParamIndex;
```

Comment:
I'm not a fan of hard-coding the kinds of parameters each method supports like this. The previous code used an enum, that made it easier to add new parameters. It also meant that things like "is this used" were just "is this enum in the list". I think we should move back to that.

I like the idea of encapsulating each method in a class. I would propose that each class object just had a list of what extra parameters were in use.